File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/web/router/routematch/RouteMatchTest.kt
package io.micronaut.docs.web.router.routematch

import io.micronaut.context.ApplicationContext
import io.micronaut.context.annotation.Requires
import io.micronaut.http.HttpRequest
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.Produces
import io.micronaut.http.client.HttpClient
import io.micronaut.runtime.server.EmbeddedServer
import io.micronaut.web.router.RouteAttributes
import io.micronaut.web.router.RouteMatch
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class RouteMatchTest {

    @Test
    fun testRouteMatchRetrieval() {
        ApplicationContext.run(EmbeddedServer::class.java, mapOf("spec.name" to "RouteMatchSpec")).use { server ->
            server.applicationContext.createBean(HttpClient::class.java, server.url).use { httpClient ->
                val request = HttpRequest.GET<Any>("/routeMatch").accept(MediaType.TEXT_PLAIN)
                assertEquals("text/plain", httpClient.toBlocking().retrieve(request, String::class.java))
            }
        }
    }
}

@Requires(property = "spec.name", value = "RouteMatchSpec")
@Controller
internal class RouteMatchController {

    @Produces(MediaType.TEXT_PLAIN)
    @Get("/routeMatch")
//tag::routematch[]
    fun index(request: HttpRequest<*>): String? {
        val routeMatch: RouteMatch<*>? = RouteAttributes.getRouteMatch(request)
            .orElse(null)
//end::routematch[]
        return routeMatch?.routeInfo?.produces?.firstOrNull()?.toString()
    }
}

File: test-suite-python/src/test/python/micronaut/docs/web/router/routematch/RouteMatchTest.py
from typing import Annotated

import java
from jakarta.inject import Inject
from micronaut.context.annotation import Property, Requires
from micronaut.http import MediaType
from micronaut.http.annotation import Controller, Get
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

HttpRequest = java.type("io.micronaut.http.HttpRequest")
RouteAttributes = java.type("io.micronaut.web.router.RouteAttributes")


@Property(name="spec.name", value="RouteMatchSpec")
@MicronautTest
class RouteMatchTest:
    client: Annotated[HttpClient, Inject, Client("/")]

    @Test
    def test_route_match_retrieval(self):
        blocking_client = self.client.toBlocking()
        request = HttpRequest.GET("/routeMatch").accept(MediaType.TEXT_PLAIN)
        assert blocking_client.retrieve(request, java.type("java.lang.String")) == "text/plain"


@Requires(property="spec.name", value="RouteMatchSpec")
@Controller
class RouteMatchController:
    @Get("/routeMatch", produces=MediaType.TEXT_PLAIN)
    # tag::routematch[]
    def index(self, request: HttpRequest) -> str:
        route_match = RouteAttributes.getRouteMatch(request).orElse(None)
    # end::routematch[]
        if route_match is None:
            return None
        produces = route_match.getRouteInfo().getProduces()
        return produces.stream().map(lambda media_type: str(media_type)).findFirst().orElse(None)
